src/constbuffer_thandle.c

```diff
@@ -0,0 +1,809 @@
+    {
+        /*Codes_SRS_CONSTBUFFER_THANDLE_88_019: [ The non-NULL handle returned by CONSTBUFFER_THANDLE_CreateWithMoveMemory shall have its ref count set to "1". ]*/
+        THANDLE(CONSTBUFFER) temp_result = THANDLE_MALLOC(CONSTBUFFER)(CONSTBUFFER_dispose);
+        if (temp_result == NULL)
+        {
+            /*Codes_SRS_CONSTBUFFER_THANDLE_88_017: [ If any error occurs, CONSTBUFFER_THANDLE_CreateWithMoveMemory shall fail and return NULL. ]*/
+            LogError("Allocation of CONSTBUFFER object failed");
+        }
+        else
+        {
+            /*Codes_SRS_CONSTBUFFER_THANDLE_88_018: [ If source is non-NULL and size is 0, the source pointer shall be owned (and freed) by the newly created instance of const buffer. ]*/
+            /*Codes_SRS_CONSTBUFFER_THANDLE_88_016: [ CONSTBUFFER_THANDLE_CreateWithMoveMemory shall store the source and size and return a non-NULL handle to the newly created const buffer. ]*/
+            CONSTBUFFER* handle_data = THANDLE_GET_T(CONSTBUFFER)(temp_result);
+            handle_data->alias.buffer = source;
+            handle_data->alias.size = size;
+            handle_data->buffer_type = CONSTBUFFER_THANDLE_TYPE_MEMORY_MOVED;
+            THANDLE_INITIALIZE(CONSTBUFFER)(&handle_data->original_handle, NULL);
+            THANDLE_MOVE(CONSTBUFFER)(&result, &temp_result);
+        }
+    }
+
+    return result;
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, THANDLE(CONSTBUFFER), CONSTBUFFER_THANDLE_CreateWithCustomFree, const unsigned char*, source, uint32_t, size, CONSTBUFFER_THANDLE_CUSTOM_FREE_FUNC, customFreeFunc, void*, customFreeFuncContext)
+{
+    THANDLE(CONSTBUFFER) result = NULL;
+
+    /*Codes_SRS_CONSTBUFFER_THANDLE_88_032: [ customFreeFuncContext shall be allowed to be NULL. ]*/
+
+    if (
+        /*Codes_SRS_CONSTBUFFER_THANDLE_88_026: [ If source is NULL and size is different than 0 then CONSTBUFFER_THANDLE_CreateWithCustomFree shall fail and return NULL. ]*/
+        ((source == NULL) && (size > 0)) ||
+        /*Codes_SRS_CONSTBUFFER_THANDLE_88_029: [ If customFreeFunc is NULL, CONSTBUFFER_THANDLE_CreateWithCustomFree shall fail and return NULL. ]*/
+        (customFreeFunc == NULL)
+        )
+    {
+        LogError("Invalid arguments: unsigned char* source=%p, uint32_t size=%" PRIu32 ", customFreeFunc=%p, customFreeFuncContext=%p",
+            source, size, customFreeFunc, customFreeFuncContext);
+    }
+    else
+    {
+        /*Codes_SRS_CONSTBUFFER_THANDLE_88_033: [ If any error occurs, CONSTBUFFER_THANDLE_CreateWithCustomFree shall fail and return NULL. ]*/
+        // Allocate extra space for the custom free function fields beyond the base structure
+        size_t extra_size = sizeof(CONSTBUFFER_THANDLE_HANDLE_WITH_CUSTOM_FREE_DATA) - sizeof(CONSTBUFFER);
+        THANDLE(CONSTBUFFER) temp_result = THANDLE_MALLOC_FLEX(CONSTBUFFER)(CONSTBUFFER_dispose, 1, extra_size);
+        if (temp_result == NULL)
+        {
+            LogError("failure in THANDLE_MALLOC_FLEX");
```

Comment:
>LogError("failure in THANDLE_MALLOC_FLEX"); [](http://example.com/codeflow?start=12&length=43)

Can we put the size of extra_size in  here.

Reply:
Done